### DISCUSSION TITLE
How can a VM message the host

### DISCUSSION THREAD

**Discussion by samwyndham** [cirruslabs/tart#337] [Q&A] [answered]
Hi everyone,

I would like to run a script on the host that is triggered by the VM. Is there a suggested way to do this?

Thank you!

**Comment by fkorotkov** [comment, marked as answer]
There is no such builtin functionality since such triggering would be a pretty big violation for contract of Virtual Machines being isolated from the host.

A common patter is when you have a unidirectional communication from the host into VM. Usually integrations SSH into VMs and execute scripts there. There might be some hacky workaround like mounting a folder from the host so a VM can write something into it. Than the host can monitor the folder and execute triggers. But it sounds a bit too much IMO.

What kind of things you'd like to trigger on the host from VMs?

**Reply by samwyndham** [reply]
Thanks @fkorotkov. That makes sense.

> What kind of things you'd like to trigger on the host from VMs?

I'm setting up some self hosted GitHub Actions runners (for a macOS project CI). I can envisage two ways of doing this:

1. The runner lives on the host and as part of the job creates/runs the VM, copies project files to VM, and builds & tests. The host then destroys the VM when the job is complete.
2. The runner is ephemeral and lives in the VM.

I'm looking into how I would implement number 2 now and considering how I would destroy the VM once the job completes. One way of achieving that is for the runner to message the host to do clean up. That is what my question relates to.

**Reply by fkorotkov** [reply]
@samwyndham we went with option 2 for [Cirrus Runners](https://github.com/apps/cirrus-runners): our orchestration spins up a VM, SSHs into it and runs an ephemeral runner. Once a runner finishes after running just one job and exits (hence `--ephemeral` flag), we delete the VM all together and start a new one if needed.
